model.py

```diff
@@ -20,31 +19,43 @@ class NNUE(pl.LightningModule):
 
   It is not ideal for training a Pytorch quantized model directly.
   """
-  def __init__(self, feature_set=halfkp.Features(), factorizer=halfkp.Factorizer(), lambda_=1.0):
+  def __init__(self, feature_set, lambda_=1.0):
     super(NNUE, self).__init__()
+    self.input = nn.Linear(feature_set.num_features, L1)
     self.feature_set = feature_set
-    self.change_factorizer(factorizer)
     self.l1 = nn.Linear(2 * L1, L2)
     self.l2 = nn.Linear(L2, L3)
     self.output = nn.Linear(L3, 1)
     self.lambda_ = lambda_
 
-  # Call to change the factorizer.  Will reset the input weights to match the expected shape.
-  def change_factorizer(self, factorizer):
-    self.factorizer = factorizer
-    num_inputs = self.feature_set.inputs
-    if self.factorizer is not None:
-      num_inputs += self.factorizer.inputs
-    # Don't reinitialize the weights if they've already been set.
-    if not hasattr(self, 'input'):
-      self.input = nn.Linear(num_inputs, L1)
-
-    if self.factorizer is not None:
-      # Zero out the weights/biases for the factorized features
-      # Weights stored as [256][41024]
-      weights = self.input.weight.narrow(1, 0, self.feature_set.inputs)
-      weights = torch.cat((weights, torch.zeros(L1, self.factorizer.inputs)), dim=1)
+    self._zero_virtual_feature_weights()
+
+  def _zero_virtual_feature_weights(self):
+    weights = self.input.weight
+    for a, b in self.feature_set.get_virtual_feature_ranges():
+      weights[a:b, :] = 0.0
+    self.input.weight = nn.Parameter(weights)
+
+  def set_feature_set(self, new_feature_set):
+    if self.feature_set.name == new_feature_set.name:
+      return
+
+    # TODO: Implement this for more complicated conversions.
+    #       Currently we support only a single feature block.
+    if len(self.feature_set.features) > 1:
+      raise Exception('Cannot change feature set from {} to {}.'.format(self.feature_set.name, new_feature_set.name))
+
+    old_feature_block = self.feature_set.features[0]
+    new_feature_block = new_feature_set.features[0]
+    if old_feature_block.name == next(iter(new_feature_block.factors)):
```

Comment:
Hm..  I can't quite follow the logic here.  Can we add a bit of description about what is happening here?